Convert BigInt/Decimal to expected types for packed numbers

jadams-tresys opened the pull request:
Still have an edge case for packed xs:date/Time values require an actual BigInt, but this fixes DAFFODIL-2961

stevedlawrence commented on daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/processors/PackedBinaryTraits.scala:
```diff
@@ -26,6 +26,8 @@ import org.apache.daffodil.lib.equality.TypeEqual
 import org.apache.daffodil.lib.exceptions.Assert
```
We should probably hold off until https://github.com/apache/daffodil/pull/1337 is merged, since that touches some of this code and is the bigger commit, easier to deal with conflicts with the smaller commit.

stevedlawrence commented on daffodil-test/src/test/scala/org/apache/daffodil/section13/packed/TestPacked.scala:
```diff
@@ -100,4 +100,7 @@ class TestPacked {
   }
+
+  // Daffodil-2961
+  @Test def testBCDBigDecimalError(): Unit = { runner.runOneTest("bcdBigDecimalError") }
```
Can you add some out of range tests as well? That's a new limitation that this will enforce that I think wasn't previously enforced.

jadams-tresys commented on daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/processors/PackedBinaryTraits.scala:
```diff
@@ -88,11 +108,8 @@ abstract class PackedBinaryIntegerBaseParser(
 
     try {
-      val int = toBigInteger(dis.getByteArray(nBits, start))
-      if (!signed && (int.signum != 1))
```
Yes, we should be able to remove that.  The error message changes to simply "Parse Error: Unable to parse xs:date from -123".  I feel that is enough of a helpful error message to indicate a non-valid time value.

jadams-tresys commented on daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/processors/PackedBinaryTraits.scala:
```diff
@@ -38,6 +40,24 @@ import passera.unsigned.ULong
+      case pn: PrimType.PrimNumeric => pn.fromNumber(toBigDecimal(num, scale))
+      // Non-numeric types such as Time can still use these funcitons and
+      // expect BigDecimal as the output of the conversion
+      case _ => toBigDecimal(num, scale)
```
I believe that is correct.  Will change.

jadams-tresys commented on daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/processors/PackedBinaryTraits.scala:
```diff
@@ -60,8 +80,8 @@ abstract class PackedBinaryDecimalBaseParser(
+      val dec = toDecimal(context, dis.getByteArray(nBits, start), binaryDecimalVirtualPoint)
+      start.simpleElement.setDataValue(dec)
     } catch {
       case n: NumberFormatException => PE(start, "Error in packed data: \n%s", n.getMessage())
```
I'll add a test to verify if this is handled gracefully or if a catch for the exception is needed.

mbeckerle commented on daffodil-test/src/test/resources/org/apache/daffodil/section13/packed/packed.tdml:
```diff
@@ -915,4 +916,37 @@
+
+  <!-- Daffodil 2961 -->
+  <tdml:parserTestCase name="bcdBigDecimalError" root="bcdBigDec" model="bcdErr"
+    description="Verify that usign packed values in an expression work correctly">
```
usign => unsigned

mbeckerle reviewed (approved):
+1
I don't have much to add beyond finding one typo.

jadams-tresys commented on daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/processors/PackedBinaryTraits.scala:
```diff
@@ -38,6 +40,24 @@ import passera.unsigned.ULong
 trait PackedBinaryConversion {
   def toBigInteger(num: Array[Byte]): JBigInteger
   def toBigDecimal(num: Array[Byte], scale: Int): JBigDecimal
+
```
The issue here is that we would still need to duplicate everything in the PackedBinaryDecimal/IntegerDelimitedBaseParsers as well.  I think that was the reason for the conversion trait to begin with.  The DelimitedBaseParsers do not share anything with the standard BaseParsers.

stevedlawrence replied:
PackedIntegerDelimitedParser only uses toBigInteger and PackedDecimalDelimitedParser only uses toBigDecimal, so these two classes do need an abstract toBigInteger/toBigDecimal respectively, but they shouldn't need both that the PackedBinaryConversion trait requires.

jadams-tresys replied:
Wouldn't the delimited parsers need the same conversion code to avoid the error this bug was fixing where a value gets used in an expression?

stevedlawrence replied:
Ah right, maybe right refactoring is more like this?

```scala
 // returns either BigDecimal or BigInteger depending on the implementation
def toBigNumber(num: Array[Byte]): Number 

def toPrimType(...) : DataValueNumber = {
  val bigNumber = toBigNumber
  context.optPrimType.get match {
    case pn: PrimType.PrimNumeric => pn.fromNumber(bigNmber)
    case => bigNumber
  }
}
```

Then all the individual parses just have to implement toBigNumber correctly, and the abstract base classes just call toPrimType. Might even be renaming it and changing it to also handle exceptions and call setDataValue to avoid more duplication.

jadams-tresys replied:
Something like that should work in theory, we just end up duplicating the toPrimType code for both the Base and Delimited parsers.  If that's preferable over having a shared trait that's fine with me.

stevedlawrence replied:
What if `PackedBinaryConversion` is changed to just contain that? Then the delimited and non-delimited parsers just call `toPrimType(arrayBytes, ...) ` and that handles everything.

jadams-tresys replied:
Just making sure that I'm on the same page.  Is that not what I did below (although they are called toInter/Decimal instead of toPrimType)?  I recognize that I need to add some exception handling for potential InvalidPrimitiveDataExceptions.  I think it also makes more sense to me to keep the calls to actually set the value inside the parse functions instead of moving them inside the toPrimType trait function.

stevedlawrence replied:
I think the main difference I'm suggesting is making the trait more generic so implementations don't have to implement both toBigInteger and toBigDecimal, they just implement a single toBigNumber and that will return the BigDecimal or BigInteger depending on if it's an decimal or integer type of parser. To ensure correct types, we could define it something like this:

```scala
trait PackedBinaryConversion[A <: Number] {
  def toNumber(num: Array[Byte]): A

  def toPrimType(...) = { ... toNumber(...)  }
}
```

And then usages would provide the right type parameter, for example:
```scala
abstract class PackedBinaryIntegerDelimitedBaseParser(...)
  ...
  with PackedBinaryConversion[JBigInteger]

...

abstract class PackedBinaryIntegerDelimitedDecimalParser(...)
  ...
  with PackedBinaryConversion[JBigDecimal]
```

Avoids duplicating the decimal/integer logic in the PackedBinaryConversion trait.

>  I think it also makes more sense to me to keep the calls to actually set the value inside the parse functions instead of moving them inside the toPrimType trait function.

I have no strong preference either way. You just have to duplicate error handling logic if it's not done in the shared trait.

jadams-tresys replied:
Is there a clean way to have a single toNumber function while also dealing with the scale parameter that is needed for Decimal?  I thought about defaulting scale to 0 and treating scale=0 as a BigInt, but scale=0 is valid for BigDecimal as well.

> You just have to duplicate error handling logic if it's not done in the shared trait.
If I have it in the trait I would then need to pass that current state in as well to create the Parse Error.  Since the PackedBinaryConversion trait I wasn't sure if there was a way to pass in either a Parse state or an Unparse state that could create errors appropriately.

I'm pushing up some changes now that I think cover everything else pretty well.

stevedlawrence replied:
Could you drop the scale parameter and get it from the class member, e.g.

```scala
class BCDDecimalKnownLengthParser(..., binaryDecimalVirtualPoint: Int)
  extends ... {

  override def toNumber(num: Array[Byte]): JBigDecimal =
    DecimalUtils.bcdToBigDecimal(num, binaryDecimalVirtualPoint)
}
```
I think this actually makes is actually an improvement over what we have now, since the abstract class shouldn't need to know anything about how the implementations converts a byte array to a `Number`, all the information like scale, sign codes, etc. is private to the specific packed binary number implementation. And if we ever need some new parameter on how a byte array is converted to a Number it can be done specifically in the implementation and the base class doesn't have to change at all.

jadams-tresys commented:
I think I've addressed all the comments from review now.  I have now rebased onto main, which now includes the fixes from #1337.  Definitely was a mistake to rebase onto that earlier before it was merged.

stevedlawrence commented on daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/processors/PackedBinaryTraits.scala:
```diff
@@ -95,24 +105,25 @@ abstract class PackedBinaryDecimalBaseParser(
   with PackedBinaryLengthCheck {
   override lazy val runtimeDependencies = Vector()
 
   val signed = {
```
This val isn't used and can be deleted? Which I think makes sense because signedness is now checked by `primType.fromNumber`?

stevedlawrence reviewed (approved):
+1 looks good :+1: , just needs lint fixes and one small change to remove a val that I think is now unused

mbeckerle commented on daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/processors/BCDParsers.scala:
```diff
@@ -83,19 +80,15 @@ class BCDIntegerRuntimeLengthParser(
-  override def toBigInteger(num: Array[Byte]): JBigInteger = DecimalUtils.bcdToBigInteger(num)
-  override def toBigDecimal(num: Array[Byte], scale: Int): JBigDecimal =
-    DecimalUtils.bcdToBigDecimal(num, scale)
+  override def toNumber(num: Array[Byte]): JBigInteger = DecimalUtils.bcdToBigInteger(num)
```
Here the return type is different. toNumber returns a JBigInteger.

I'm surprised this even works. JBigInteger must be a sub-type of JBigDecimal.

mbeckerle commented on daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/processors/BCDParsers.scala:
```diff
@@ -34,9 +34,8 @@ class BCDDecimalKnownLengthParser(
-  override def toBigInteger(num: Array[Byte]): JBigInteger = DecimalUtils.bcdToBigInteger(num)
-  override def toBigDecimal(num: Array[Byte], scale: Int): JBigDecimal =
-    DecimalUtils.bcdToBigDecimal(num, scale)
+  override def toNumber(num: Array[Byte]): JBigDecimal =
```
Why change this to toNumber? The return type is not JNumber. 

Is there scaladoc on the toNumber method? There needs to be to explain this naming convention. I don't really understand why this got renamed.

stevedlawrence replied:
The trait these classes mix in requires an implementation of `def toNumber(num: Array[Byte]): A`, where `A <: Number`. We could remove the `A` type parameter and just make it so it always returns a `JNumber`, but then implementations have to be a bit more careful to return the right thing.

As it is right now, the type parameter ensures the implementations of `toNumber` return the right kind of `JNumber` (some should always return a BigInteger, some should always return a BigDecimal). I'm not sure the type parameter is used anywhere else though, so maybe that's overkill and just adds unnecessary complication?

mbeckerle replied:
I think the type parameter is almost certainly the right thing, I just don't see a change that adds this toNumber super method declaration, yet it's not in the current Daffodil main branch, so when did this type parameter get added? 

But what is the purpose of the toNumber method? I think it is "convert to a number type we can store in the Infoset object. This should be the most specific meaningful subtype of JNumber" that is used in the Infoset object representation. Or something like that.

jadams-tresys replied:
The top level declaration of toNumber is in PackedBinaryTraits.scala where it is simply declared as
`trait PackedBinaryConversion[A <: Number] {
    def toNumber(num: Array[Byte]): A
    ....`

The actual implementations are in the classes that extend the PackedBinaryConversion trait, like BCDIntegerKnownLengthParser.

The function will only generate either BigInteger's or BigDecimal's depending on the calling classes type, so perhaps a better name for the function would be "toBigNumber".  The BigNumber then gets converted to the appropriate PrimType by calling PrimNumberic.fromNumber(bigNumber).

mbeckerle reviewed (approved):
Just added some questions, and suggest adding a scaladoc to motivate the toNumber method name change.

jadams-tresys commented:
I've made another small change that I had forgotten about.  We don't want to check the bit length of binary numbers during schema compilation if the binaryNumberRep is not "binary" as it can take more than 4 bytes to represent an xs:int for example.  Please double check that this simple extra conditional is a reasonable solution to this.

stevedlawrence commented on daffodil-test/src/test/resources/org/apache/daffodil/section13/packed/packed.tdml:
```diff
@@ -1019,6 +1020,20 @@
+      <tdml:error>Error in packed data</tdml:error>
+      <tdml:error>out of range for type</tdml:error>
+    </tdml:errors>
+    </tdml:parserTestCase>
```
Can we add a positive test case for this? To show that you can have a large amount of bytes for packed values as long as the result is within range of the type.

jadams-tresys replied:
Done

stevedlawrence commented on daffodil-test/src/test/resources/org/apache/daffodil/section13/packed/packed.tdml:
```diff
@@ -965,4 +966,88 @@
+    </tdml:errors>
+    </tdml:parserTestCase>
+
+  <tdml:parserTestCase name="packedIntMax" root="int01" model="packedDecimalRangeChecks"
+    description="Using BCD formatted hex to indicate INT_MAX + 1">
```
It doesn't look like you added a junit test in the scala file for this test.

Also, I think the description should say INT_MAX instead of INT_MAX + 1.

stevedlawrence reviewed (approved):
+1 to the latest changes with some minor fixes
